pkg/policy/types/types.go

```diff
@@ -100,86 +87,61 @@ func (k Key) IsEgress() bool {
 
 // EndPort returns the end-port of the Key based on the Mask.
 func (k Key) EndPort() uint16 {
-	return k.DestPort + k.invertedPortMask
+	return k.DestPort + uint16(0xffff)>>k.PortPrefixLen()
 }
 
 // PortProtoIsBroader returns true if the receiver Key has broader
 // port-protocol than the argument Key. That is a port-protocol
 // that covers the argument Key's port-protocol and is larger.
 // An equal port-protocol will return false.
 func (k Key) PortProtoIsBroader(c Key) bool {
-	if k.Nexthdr == 0 && c.Nexthdr != 0 {
-		return k.PortIsEqual(c) || k.PortIsBroader(c)
-	}
-	return k.Nexthdr == c.Nexthdr && k.PortIsBroader(c)
+	// Port is wildcarded when protocol is wildcarded
+	return k.Nexthdr == 0 && c.Nexthdr != 0 || k.Nexthdr == c.Nexthdr && k.PortIsBroader(c)
 }
 
 // PortProtoIsEqual returns true if the port-protocols of the
 // two keys are exactly equal.
 func (k Key) PortProtoIsEqual(c Key) bool {
-	return k.DestPort == c.DestPort &&
-		k.invertedPortMask == c.invertedPortMask &&
-		k.Nexthdr == c.Nexthdr
+	return k.Nexthdr == c.Nexthdr && k.PortIsEqual(c)
 }
 
 // PortIsBroader returns true if the receiver Key's
 // port range covers the argument Key's port range,
 // but returns false if they are equal.
 func (k Key) PortIsBroader(c Key) bool {
-	if k.DestPort == 0 && c.DestPort != 0 {
-		return true
-	}
-	if k.DestPort != 0 && c.DestPort == 0 {
-		return false
-	}
-	kEP := k.EndPort()
-	cEP := c.EndPort()
-	return k.DestPort <= c.DestPort && kEP >= cEP &&
-		// The port ranges cannot be exactly equal.
-		(k.DestPort != c.DestPort || kEP != cEP)
+	// Broader port must have shorter prefix and a common part needs to be the same
+	kPrefixLen := k.PortPrefixLen()
+	cPrefixLen := c.PortPrefixLen()
+	return kPrefixLen < cPrefixLen &&
+		k.DestPort^c.DestPort&(uint16(0xffff)<<(16-kPrefixLen)) == 0
 }
 
 // PortIsEqual returns true if the port ranges
 // between the two keys are exactly equal.
 func (k Key) PortIsEqual(c Key) bool {
-	return k.DestPort == c.DestPort &&
-		k.invertedPortMask == c.invertedPortMask
+	return k.DestPort == c.DestPort && k.bits<<1 == c.bits<<1 // ignore traffic direction
 }
 
 // PrefixLength returns the prefix lenth of the key
 // for indexing it for the userspace cache (not the
 // BPF map or datapath).
 func (k Key) PrefixLength() uint {
-	keyPrefix := MapStatePrefixLen
-	portPrefix := uint(16)
-	if k.DestPort != 0 {
-		// It is not possible for k.InvertedPortMask
-		// to be incorrectly set, but even if
-		// it was the default value of "0" is
-		// what we want.
-		portPrefix = uint(bits.TrailingZeros16(k.PortMask()))
-	}
-	keyPrefix -= portPrefix
-	// If the port is fully wildcarded then
-	// we can also wildcard the protocol
-	// (if it is also wildcarded).
-	if portPrefix == 16 && k.Nexthdr == 0 {
-		keyPrefix -= 8
+	if k.Nexthdr == 0 {
+		return 1 // direction always specified
 	}
-	return keyPrefix
+	return 9 + uint(k.PortPrefixLen())
```

Comment:
Can you comment what this number is for?

Reply:
Added comment:
```
	// 1 bit from direction bit + 8 bits for the protocol when `k.Nexthdr' != 0
```